vnext/ReactWindowsCore/tracing/fbsystrace.h

```diff
@@ -51,44 +55,48 @@ class FbSystraceSection {
             .count());
   }
 
-  void init(std::string &&v) {
-    args_[index_++] = std::move(v);
-    begin_section();
+  template <typename... RestArg>
+  FbSystraceSection(uint64_t tag, std::string &&profileName, RestArg &&... rest)
+      : tag_(tag), profile_name_(std::move(profileName)) {
+    id_ = s_id_counter++;
+    init(std::forward<RestArg>(rest)...);
   }
 
-  void init(const std::string &v) {
-    args_[index_++] = v; // copy
-    begin_section();
+  ~FbSystraceSection() {
+    end_section();
   }
 
-  void init() {
-    begin_section();
+ private:
+  template <typename T>
+  using ConsructsToString =
+      std::bool_constant<std::is_convertible<T, std::string>::value || std::is_constructible<std::string, T>::value>;
+
+  template <typename Arg, typename std::enable_if_t<ConsructsToString<Arg>::value> * = nullptr>
+  inline void push_arg(Arg &&arg) {
+    args_[index_++] = std::forward<Arg>(arg);
   }
 
-  template <typename... ConvertsToStringPiece>
-  void init(std::string &&v, ConvertsToStringPiece &&... rest) {
-    args_[index_++] = std::move(v);
-    init(std::forward<ConvertsToStringPiece>(rest)...);
+  template <typename Arg, typename std::enable_if_t<!ConsructsToString<Arg>::value> * = nullptr>
+  inline void push_arg(Arg &&arg) {
```

Comment:
>inline void push_arg(Arg &&arg) { [](start = 2, length = 33)

It all looks too complex: I would suggest to remove these overloaded methods and instead in the init() method below that takes variadic args to use 'if constexpr()' and check the value type and do conditional conversion there. It would be the same effect with less code.

Reply:
You're right. `constexpr if` with a type check ends up being a lot cleaner than SFINAE. Will change this.